File: SGUEES-API/sguees.api/Options/SelectionHiring/SC_MOVIMIENTO_PERSONAL/SC_MOVIMIENTO_PERSONALService.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using SGUEES.Models;
using SGUEES.Repositories;

namespace SGUEES.Services
{
	/// <summary>
	/// Reglas de negocio del movimiento de personal (CRUD + Autoriza + lookups).
	/// </summary>
	public class SC_MOVIMIENTO_PERSONALService : ISC_MOVIMIENTO_PERSONALService
	{
		private readonly ISC_MOVIMIENTO_PERSONALRepository _repo;

		public SC_MOVIMIENTO_PERSONALService(ISC_MOVIMIENTO_PERSONALRepository repo)
		{
			_repo = repo;
		}

		public async Task<CResult> GetAllAsync(SC_MOVIMIENTO_PERSONALParam xWhere)
		{
			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetAllAsync(p);
		}

		public async Task<CResult> GetAsync(SC_MOVIMIENTO_PERSONALParam xWhere)
		{
			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
				new() { ParameterName = "CORR_MOVIMIENTO_PERSONAL", Value = xWhere.CORR_MOVIMIENTO_PERSONAL, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetAsync(p);
		}

		public async Task<CResult> CreateAsync(SC_MOVIMIENTO_PERSONALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			var validacion = ValidarNegocio(Data, esAlta: true);
			if (validacion != null)
			{
				return validacion;
			}

			NormalizarAlta(Data);
			return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> UpdateAsync(SC_MOVIMIENTO_PERSONALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			/* Estado real desde BD (no confiar en el body del cliente). */
			var actual = await GetAsync(new SC_MOVIMIENTO_PERSONALParam
			{
				CORR_EMPRESA = Data.CORR_EMPRESA,
				CORR_MOVIMIENTO_PERSONAL = Data.CORR_MOVIMIENTO_PERSONAL,
			});

			if (actual.Data is not SC_MOVIMIENTO_PERSONALView row)
			{
				return ValidationError("No se encontró el movimiento a modificar.");
			}

			Data.ESTADO_MOVIMIENTO = row.ESTADO_MOVIMIENTO;
			Data.ORIGEN_MOVIMIENTO = row.ORIGEN_MOVIMIENTO;

			var validacion = ValidarNegocio(Data, esAlta: false);
			if (validacion != null)
			{
				return validacion;
			}

			return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> DeleteAsync(SC_MOVIMIENTO_PERSONALTable Data, string vLOGIN_SISTEMA, string vESTACION)
		{
			/* Solo se elimina en Borrador. */
			var actual = await GetAsync(new SC_MOVIMIENTO_PERSONALParam
			{
				CORR_EMPRESA = Data.CORR_EMPRESA,
				CORR_MOVIMIENTO_PERSONAL = Data.CORR_MOVIMIENTO_PERSONAL,
			});

			if (actual.Data is SC_MOVIMIENTO_PERSONALView row
				&& !string.Equals(row.ESTADO_MOVIMIENTO, "DI", System.StringComparison.OrdinalIgnoreCase))
			{
				return ValidationError("Solo se puede eliminar un movimiento en estado Borrador.");
			}

			return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
		}

		public async Task<CResult> AutorizaAsync(SC_MOVIMIENTO_PERSONAL_AUTORIZAParam Data, string vLOGIN_SISTEMA)
		{
			if (Data == null)
			{
				return ValidationError("No se recibieron datos para autorizar el movimiento.");
			}

			if (Data.CORR_EMPRESA <= 0)
			{
				return ValidationError("No se pudo identificar la empresa de la sesión.");
			}

			if (Data.CORR_MOVIMIENTO_PERSONAL <= 0)
			{
				return ValidationError("Debe indicar el movimiento de personal.");
			}

			if (Data.OPERACION < 1 || Data.OPERACION > 5)
			{
				return ValidationError("Operación inválida. Use 1=GUARDAR, 2=ENVIAR, 3=APROBAR, 4=DEVOLVER, 5=RECHAZAR.");
			}

			if (string.IsNullOrWhiteSpace(Data.OBSERVACION))
			{
				return ValidationError("El comentario / observación es obligatorio.");
			}

			if (string.IsNullOrWhiteSpace(vLOGIN_SISTEMA))
			{
				return ValidationError("No se pudo identificar el usuario de sesión.");
			}

			Data.OBSERVACION = Data.OBSERVACION.Trim();
			return await _repo.AutorizaAsync(Data, vLOGIN_SISTEMA.Trim());
		}

		public async Task<CResult> GetBitacoraAsync(SC_MOVIMIENTO_PERSONAL_BITACORAParam xWhere)
		{
			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_TIPO_DOCUMENTO", Value = xWhere.CORR_TIPO_DOCUMENTO, DbType = System.Data.DbType.Int32 },
				new() { ParameterName = "CORR_DOCUMENTO", Value = xWhere.CORR_MOVIMIENTO_PERSONAL, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetAllAsyncBitacora(p);
		}

		public async Task<CResult> GetUnidadesUsuarioAsync(SC_MOVIMIENTO_PERSONALParam xWhere)
		{
			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
				new() { ParameterName = "LOGIN_SISTEMA", Value = xWhere.LOGIN_SISTEMA ?? string.Empty, DbType = System.Data.DbType.String },
			};

			return await _repo.GetUnidadesUsuarioAsync(p);
		}

		public async Task<CResult> GetPuestosByUnidadAsync(SC_MOVIMIENTO_PERSONALParam xWhere)
		{
			if (xWhere.CORR_UNIDAD <= 0)
			{
				return new CResult
				{
					Data = new List<SC_MOVIMIENTO_LOOKUP_PUESTOView>(),
					Result = true,
					RowsAffected = 0,
					ErrorCode = 0,
				};
			}

			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
				new() { ParameterName = "CORR_UNIDAD", Value = xWhere.CORR_UNIDAD, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetPuestosByUnidadAsync(p);
		}

		public async Task<CResult> GetModalidadesAsync(SC_MOVIMIENTO_PERSONALParam xWhere)
		{
			var p = new List<CParameter>
			{
				new() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
			};

			return await _repo.GetModalidadesAsync(p);
		}

		/// <summary>Defaults al crear: origen DIRECTO, estado DI.</summary>
		private static void NormalizarAlta(SC_MOVIMIENTO_PERSONALTable Data)
		{
			if (string.IsNullOrWhiteSpace(Data.ORIGEN_MOVIMIENTO))
			{
				Data.ORIGEN_MOVIMIENTO = "DIRECTO";
			}

			Data.ORIGEN_MOVIMIENTO = Data.ORIGEN_MOVIMIENTO.Trim().ToUpperInvariant();
			Data.TIPO_MOVIMIENTO = (Data.TIPO_MOVIMIENTO ?? "PERMANENTE").Trim().ToUpperInvariant();

			/* Desde requisición nace aprobado; desde cero nace en borrador. */
			if (Data.ORIGEN_MOVIMIENTO == "REQUISICION")
			{
				Data.ESTADO_MOVIMIENTO = "AP";
				if (!Data.FECHA_EFECTIVA.HasValue)
				{
					Data.FECHA_EFECTIVA = System.DateTime.Today;
				}
			}
			else
			{
				Data.ESTADO_MOVIMIENTO = "DI";
			}

			if (Data.FECHA_ELABORACION.Year < 1753)
			{
				Data.FECHA_ELABORACION = System.DateTime.Today;
			}
		}

		private static CResult ValidarNegocio(SC_MOVIMIENTO_PERSONALTable Data, bool esAlta)
		{
			if (Data == null)
			{
				return ValidationError("No se recibieron datos del movimiento.");
			}

			if (string.IsNullOrWhiteSpace(Data.NOMBRE_COMPLETO))
			{
				return ValidationError("Debe indicar el nombre completo.");
			}

			var tipo = (Data.TIPO_MOVIMIENTO ?? string.Empty).Trim().ToUpperInvariant();
			if (tipo is not ("PERMANENTE" or "EVENTUAL" or "ASCENSO" or "TRASLADO"))
			{
				return ValidationError("Tipo de movimiento inválido. Use PERMANENTE, EVENTUAL, ASCENSO o TRASLADO.");
			}

			if (!Data.CORR_UNIDAD_PROPUESTA.HasValue || Data.CORR_UNIDAD_PROPUESTA.Value <= 0)
			{
				return ValidationError("Debe indicar la unidad / departamento propuesto.");
			}

			if (!Data.CORR_PUESTO_PROPUESTO.HasValue || Data.CORR_PUESTO_PROPUESTO.Value <= 0)
			{
				return ValidationError("Debe indicar el cargo / puesto propuesto.");
			}

			if (Data.SALARIO_PROPUESTO.HasValue && Data.SALARIO_PROPUESTO.Value < 0)
			{
				return ValidationError("El salario propuesto no puede ser negativo.");
			}

			if (tipo == "EVENTUAL")
			{
				if (!Data.FECHA_FINALIZACION.HasValue)
				{
					return ValidationError("En contratación eventual debe indicar la fecha de finalización.");
				}

				if (Data.FECHA_INGRESO_PROPUESTA.HasValue
					&& Data.FECHA_FINALIZACION.Value.Date < Data.FECHA_INGRESO_PROPUESTA.Value.Date)
				{
					return ValidationError("La fecha de finalización no puede ser anterior a la de ingreso propuesta.");
				}
			}

			if (tipo is "ASCENSO" or "TRASLADO")
			{
				if (!Data.CORR_UNIDAD_ACTUAL.HasValue || Data.CORR_UNIDAD_ACTUAL.Value <= 0
					|| !Data.CORR_PUESTO_ACTUAL.HasValue || Data.CORR_PUESTO_ACTUAL.Value <= 0)
				{
					return ValidationError("En ascenso o traslado debe indicar la posición actual (unidad y puesto).");
				}
			}

			if (!esAlta)
			{
				var estado = (Data.ESTADO_MOVIMIENTO ?? "DI").Trim().ToUpperInvariant();
				if (estado is not ("DI" or "OB"))
				{
					return ValidationError("Solo se puede modificar un movimiento en Borrador o Devuelto.");
				}
			}

			return null;
		}

		private static CResult ValidationError(string message)
		{
			return new CResult
			{
				Data = null,
				Result = false,
				RowsAffected = 0,
				CodeHelper = 0,
				ErrorCode = 4101,
				ErrorMessage = message,
				ErrorSource = "SC_MOVIMIENTO_PERSONALService",
			};
		}
	}
}

File: SGUEES-SPA/src/app/pages/SelectionHiring/sc-movimiento-personal/sc-movimiento-personal.service.ts
import { Observable } from 'rxjs';
import { Injectable } from '@angular/core';
import { IResult } from 'src/app/FxAPI/IResult';
import { NotifyType } from 'src/app/shared/models/NotifyType';

import { ScMovimientoPersonalRepository } from './sc-movimiento-personal.repository';
import { ScMovimientoPersonal } from './models/sc-movimiento-personal';

@Injectable({
	providedIn: 'root',
})
export class ScMovimientoPersonalService {
	constructor(private repo: ScMovimientoPersonalRepository) {}

	/** Catálogo de estados de negocio (chip / grid). */
	readonly estadosMovimiento: { CODIGO: string; NOMBRE: string }[] = [
		{ CODIGO: 'DI', NOMBRE: 'Borrador' },
		{ CODIGO: 'SO', NOMBRE: 'Solicitado' },
		{ CODIGO: 'OB', NOMBRE: 'Devuelto' },
		{ CODIGO: 'AP', NOMBRE: 'Aprobado' },
		{ CODIGO: 'DE', NOMBRE: 'Denegado' },
		{ CODIGO: 'AN', NOMBRE: 'Anulado' },
	];

	readonly tiposMovimiento: { CODIGO: string; NOMBRE: string }[] = [
		//{ CODIGO: 'PERMANENTE', NOMBRE: 'Contratación permanente' },
		//{ CODIGO: 'EVENTUAL', NOMBRE: 'Contratación eventual' },
		{ CODIGO: 'ASCENSO', NOMBRE: 'Ascenso' },
		{ CODIGO: 'TRASLADO', NOMBRE: 'Traslado' },
	];

	//#region Validadores
	esValido(model: ScMovimientoPersonal, msg: Function): boolean {
		const origen = `${model?.ORIGEN_MOVIMIENTO || 'DIRECTO'}`.trim().toUpperCase();
		if (origen === 'DIRECTO' && !(Number(model?.CORR_EMPLEADO) > 0)) {
			msg('Debe seleccionar el empleado.', NotifyType.Warning);
			return false;
		}

		if (!model?.NOMBRE_COMPLETO?.trim()) {
			msg('Debe indicar el nombre completo.', NotifyType.Warning);
			return false;
		}

		const tipo = `${model.TIPO_MOVIMIENTO || ''}`.trim().toUpperCase();
		if (!['PERMANENTE', 'EVENTUAL', 'ASCENSO', 'TRASLADO'].includes(tipo)) {
			msg('Debe seleccionar un tipo de movimiento.', NotifyType.Warning);
			return false;
		}

		if (!`${model.GERENCIA_PROPUESTA || ''}`.trim()) {
			msg('Debe indicar la gerencia / vicerrectoría / facultad propuesta.', NotifyType.Warning);
			return false;
		}

		if (!model.CORR_UNIDAD_PROPUESTA || model.CORR_UNIDAD_PROPUESTA <= 0) {
			msg('Debe indicar la unidad / departamento propuesto.', NotifyType.Warning);
			return false;
		}

		if (!model.CORR_PUESTO_PROPUESTO || model.CORR_PUESTO_PROPUESTO <= 0) {
			msg('Debe indicar el cargo propuesto.', NotifyType.Warning);
			return false;
		}

		if (model.SALARIO_PROPUESTO != null && Number(model.SALARIO_PROPUESTO) < 0) {
			msg('El salario propuesto no puede ser negativo.', NotifyType.Warning);
			return false;
		}

		if (this.muestraFechaFinalizacion(model)) {
			if (!model.FECHA_FINALIZACION) {
				msg('En contratación eventual debe indicar la fecha de finalización.', NotifyType.Warning);
				return false;
			}
			if (
				model.FECHA_INGRESO_PROPUESTA &&
				new Date(model.FECHA_FINALIZACION) < new Date(model.FECHA_INGRESO_PROPUESTA)
			) {
				msg('La fecha de finalización no puede ser anterior a la de ingreso.', NotifyType.Warning);
				return false;
			}
		} else {
			model.FECHA_FINALIZACION = null;
		}

		if (tipo === 'ASCENSO' || tipo === 'TRASLADO') {
			if (!model.CORR_UNIDAD_ACTUAL || model.CORR_UNIDAD_ACTUAL <= 0
				|| !model.CORR_PUESTO_ACTUAL || model.CORR_PUESTO_ACTUAL <= 0) {
				msg('En ascenso o traslado debe indicar la posición actual.', NotifyType.Warning);
				return false;
			}
		}

		return true;
	}

	/**
	 * Fecha de finalización solo si el movimiento viene de requisición
	 * y esa requisición es contratación eventual (CORR_TIPO_CONTRATACION = 2).
	 */
	muestraFechaFinalizacion(model?: Partial<ScMovimientoPersonal> | null): boolean {
		const origen = `${model?.ORIGEN_MOVIMIENTO || ''}`.trim().toUpperCase();
		return origen === 'REQUISICION' && Number(model?.CORR_TIPO_CONTRATACION) === 2;
	}

	/** Editable solo Borrador (DI) o Devuelto (OB). */
	esEstadoEditable(estado?: string): boolean {
		const e = `${estado || 'DI'}`.trim().toUpperCase();
		return e === 'DI' || e === 'OB';
	}

	/** Enviable: mismos estados editables. */
	esEstadoEnviable(estado?: string): boolean {
		return this.esEstadoEditable(estado);
	}

	/**
	 * Confirmable: CONFIRMADO=0 y
	 * (REQUISICION desde que existe) o (DIRECTO solo si ESTADO=AP).
	 */
	esConfirmable(model?: Partial<ScMovimientoPersonal> | null): boolean {
		if (!model || (Number(model.CORR_MOVIMIENTO_PERSONAL) || 0) <= 0) {
			return false;
		}
		if (this.esConfirmado(model.CONFIRMADO)) {
			return false;
		}

		const origen = `${model.ORIGEN_MOVIMIENTO || ''}`.trim().toUpperCase();
		const estado = `${model.ESTADO_MOVIMIENTO || ''}`.trim().toUpperCase();

		if (origen === 'REQUISICION') {
			return true;
		}
		if (origen === 'DIRECTO') {
			return estado === 'AP';
		}
		return false;
	}

	esConfirmado(valor?: boolean | number | null): boolean {
		return valor === true || valor === 1;
	}

	getConfirmacionLabel(model?: Partial<ScMovimientoPersonal> | null): string {
		if (model?.NOMBRE_CONFIRMACION) {
			return model.NOMBRE_CONFIRMACION;
		}
		return this.esConfirmado(model?.CONFIRMADO) ? 'Confirmado' : 'En Evaluación';
	}

	getEstadoLabel(estado?: string): string {
		const e = `${estado || 'DI'}`.trim().toUpperCase();
		return this.estadosMovimiento.find((x) => x.CODIGO === e)?.NOMBRE ?? e;
	}

	getEstadoBadgeClass(estado?: string): string {
		switch (`${estado || 'DI'}`.trim().toUpperCase()) {
			case 'DI':
				return 'estado-mov--borrador';
			case 'SO':
				return 'estado-mov--solicitado';
			case 'OB':
				return 'estado-mov--devuelto';
			case 'AP':
				return 'estado-mov--aprobado';
			case 'DE':
				return 'estado-mov--denegado';
			case 'AN':
				return 'estado-mov--anulado';
			default:
				return 'estado-mov--borrador';
		}
	}
	//#endregion

	//#region CRUD
	getAll(param: any): Observable<IResult> {
		return this.repo.getAll([{ Parameter: 'CORR_MOVIMIENTO_PERSONAL', Value: param.CORR_MOVIMIENTO_PERSONAL ?? 0 }]);
	}

	insert(model: any): Observable<IResult> {
		return this.repo.create(model);
	}

	update(model: any): Observable<IResult> {
		return this.repo.update(model, [
			{ Parameter: 'CORR_MOVIMIENTO_PERSONAL', Value: model.CORR_MOVIMIENTO_PERSONAL },
		]);
	}

	delete(param: any): Observable<IResult> {
		return this.repo.delete([
			{ Parameter: 'CORR_MOVIMIENTO_PERSONAL', Value: param.CORR_MOVIMIENTO_PERSONAL },
		]);
	}

	autoriza(model: {
		CORR_MOVIMIENTO_PERSONAL: number;
		OPERACION: number;
		OBSERVACION: string;
		CORR_UNIDAD_DOCUMENTO?: number | null;
	}): Observable<IResult> {
		return this.repo.autoriza(model);
	}

	confirmar(model: {
		CORR_MOVIMIENTO_PERSONAL: number;
		FECHA_EFECTIVA?: Date | string | null;
	}): Observable<IResult> {
		return this.repo.confirmar(model);
	}

	getBitacora(param: any): Observable<IResult> {
		return this.repo.getBitacora([
			{ Parameter: 'CORR_MOVIMIENTO_PERSONAL', Value: param.CORR_MOVIMIENTO_PERSONAL },
		]);
	}
	//#endregion

	//#region Grid / Form
	getColumns(): any[] {
		return [
			{ dataField: 'CORR_MOVIMIENTO_PERSONAL', caption: 'Corr.', width: 90 },
			{ dataField: 'FECHA_ELABORACION', caption: 'Fecha', width: 120, dataType: 'date', format: 'dd/MM/yyyy' },
			{
				dataField: 'ESTADO_MOVIMIENTO',
				caption: 'Estado',
				width: 130,
				alignment: 'center',
				allowFiltering: true,
				calculateCellValue: (row: any) => this.getEstadoLabel(row?.ESTADO_MOVIMIENTO),
				cellTemplate: (container: HTMLElement, options: any) => {
					const estado = options?.data?.ESTADO_MOVIMIENTO;
					const chip = document.createElement('span');
					chip.className = `estado-mov-chip ${this.getEstadoBadgeClass(estado)}`;
					chip.textContent = this.getEstadoLabel(estado);
					chip.title = chip.textContent;
					container.appendChild(chip);
				},
			},
			{
				dataField: 'NOMBRE_CONFIRMACION',
				caption: 'Confirmación',
				width: 150,
				alignment: 'center',
				calculateCellValue: (row: any) => this.getConfirmacionLabel(row),
				cellTemplate: (container: HTMLElement, options: any) => {
					const row = options?.data;
					const confirmado = this.esConfirmado(row?.CONFIRMADO);
					const chip = document.createElement('span');
					chip.className = `estado-mov-chip ${confirmado ? 'estado-mov--confirmado' : 'estado-mov--evaluacion'}`;
					chip.textContent = this.getConfirmacionLabel(row);
					chip.title = chip.textContent;
					container.appendChild(chip);
				},
			},
			{ dataField: 'NOMBRE_TIPO_MOVIMIENTO', caption: 'Tipo', width: 100 },
			{ dataField: 'NOMBRE_COMPLETO', caption: 'Nombre', width: 250 },
			{ dataField: 'NUMERO_ID', caption: 'Documento', width: 130 },
			{ dataField: 'NOMBRE_UNIDAD_PROPUESTA', caption: 'Unidad propuesta', width: 200 },
			{ dataField: 'NOMBRE_PUESTO_PROPUESTO', caption: 'Puesto propuesto', width: 200 },
			{ dataField: 'SALARIO_PROPUESTO', caption: 'Salario prop.', width: 140, format: '#,##0.00' },
			{ dataField: 'NOMBRE_ORIGEN_MOVIMIENTO', caption: 'Origen', width: 140 },
			{ dataField: 'USUARIO_CREA', caption: 'Usuario crea', width: 140 },
			{ dataField: 'FECHA_CREA', caption: 'Fecha crea', width: 140, dataType: 'date', format: 'dd/MM/yyyy' },
		];
	}

	getSummary(): any {
		return {
			totalItems: [
				{
					column: 'CORR_MOVIMIENTO_PERSONAL',
					summaryType: 'count',
					valueFormat: '#,##0',
					displayFormat: 'Cant: {0}',
				},
			],
		};
	}

	getBitacoraColumns(): any[] {
		return [
			{ dataField: 'FECHA_ACCION', caption: 'Fecha', width: 160, dataType: 'datetime', format: 'dd/MM/yyyy HH:mm' },
			{ dataField: 'LOGIN_SISTEMA', caption: 'Usuario', width: 140 },
			{ dataField: 'ESTADO_DESTINO', caption: 'Estado', width: 160 },
			{ dataField: 'COMENTARIO', caption: 'Comentario', width: 400 },
		];
	}

	getItems(origenMovimiento: string = 'DIRECTO', corrTipoContratacion?: number | null): any[] {
		const esDirecto = `${origenMovimiento || 'DIRECTO'}`.trim().toUpperCase() === 'DIRECTO';
		const muestraFechaFinalizacion = this.muestraFechaFinalizacion({
			ORIGEN_MOVIMIENTO: origenMovimiento,
			CORR_TIPO_CONTRATACION: corrTipoContratacion,
		});
		const campoFechaFinalizacion = muestraFechaFinalizacion
			? {
					dataField: 'FECHA_FINALIZACION',
					label: { text: 'Fecha de finalización (eventual)' },
					colSpan: 2,
					editorType: 'dxDateBox',
					editorOptions: { displayFormat: 'dd/MM/yyyy', openOnFieldClick: true, showClearButton: true },
			  }
			: null;
		const campoPersona = esDirecto
			? {
					dataField: 'CORR_EMPLEADO',
					label: { text: 'Empleado' },
					colSpan: 8,
					template: 'CORR_EMPLEADOLookup',
			  }
			: {
					dataField: 'NOMBRE_COMPLETO',
					label: { text: 'Nombre completo' },
					colSpan: 4,
					editorOptions: { maxLength: 250, showClearButton: true },
			  };

		return [
			{
				itemType: 'group',
				colCount: 8,
				colSpan: 8,
				cssClass: 'movimiento-grupo-encabezado',
				items: [
					{
						itemType: 'simple',
						colSpan: 8,
						template: 'EncabezadoConFechaTemplate',
						cssClass: 'movimiento-encabezado-titulo-item',
					},
					{
						dataField: 'CORR_MOVIMIENTO_PERSONAL',
						label: { text: 'Corr.' },
						colSpan: 1,
						editorOptions: { readOnly: true },
					},
					{
						dataField: 'ESTADO_MOVIMIENTO',
						label: { text: 'Estado' },
						colSpan: 1,
						template: 'ESTADO_MOVIMIENTOChip',
					},
					{
						dataField: 'ORIGEN_MOVIMIENTO',
						label: { text: 'Origen' },
						colSpan: 2,
						editorType: 'dxSelectBox',
						editorOptions: {
							dataSource: [
								{ CODIGO: 'DIRECTO', NOMBRE: 'Desde cero' },
								{ CODIGO: 'REQUISICION', NOMBRE: 'Desde requisición' },
							],
							valueExpr: 'CODIGO',
							displayExpr: 'NOMBRE',
							readOnly: true,
						},
					},
					{
						dataField: 'FECHA_ELABORACION',
						label: { text: 'Fecha de elaboración' },
						colSpan: 2,
						editorType: 'dxDateBox',
						editorOptions: { displayFormat: 'dd/MM/yyyy', openOnFieldClick: true },
					},
					{
						dataField: 'TIPO_MOVIMIENTO',
						label: { text: 'Tipo de movimiento' },
						colSpan: 2,
						editorType: 'dxSelectBox',
						editorOptions: {
							dataSource: this.tiposMovimiento,
							valueExpr: 'CODIGO',
							displayExpr: 'NOMBRE',
							searchEnabled: false,
						},
					},
					campoPersona,
					{
						dataField: 'NUMERO_ID',
						label: { text: 'Número de ID / documento' },
						colSpan: 2,
						editorOptions: { maxLength: 50, showClearButton: true },
					},
					{
						dataField: 'FECHA_INGRESO_PROPUESTA',
						label: { text: 'Fecha de ingreso propuesta' },
						colSpan: 2,
						editorType: 'dxDateBox',
						editorOptions: { displayFormat: 'dd/MM/yyyy', openOnFieldClick: true, showClearButton: true },
					},
					...(campoFechaFinalizacion ? [campoFechaFinalizacion] : []),
				],
			},
			{
				itemType: 'group',
				caption: 'Posición actual',
				colCount: 8,
				colSpan: 8,
				items: [
					{
						dataField: 'GERENCIA_ACTUAL',
						label: { text: 'Gerencia / Vicerrectoría / Facultad' },
						colSpan: 4,
						editorOptions: { maxLength: 200, showClearButton: true },
					},
					{
						dataField: 'CORR_UNIDAD_ACTUAL',
						label: { text: 'Departamento / Unidad' },
						colSpan: 4,
						template: 'CORR_UNIDAD_ACTUALLookup',
					},
					{
						dataField: 'CORR_PUESTO_ACTUAL',
						label: { text: 'Cargo' },
						colSpan: 4,
						template: 'CORR_PUESTO_ACTUALLookup',
					},
					{
						dataField: 'SALARIO_ACTUAL',
						label: { text: 'Salario mensual' },
						colSpan: 2,
						editorType: 'dxNumberBox',
						editorOptions: { format: '#,##0.00', min: 0, showClearButton: true },
					},
					{
						dataField: 'CORR_TIPO_MODALIDAD_ACTUAL',
						label: { text: 'Modalidad de trabajo' },
						colSpan: 2,
						template: 'CORR_TIPO_MODALIDAD_ACTUALLookup',
					},
					{
						dataField: 'HORARIO_ACTUAL',
						label: { text: 'Horario de trabajo' },
						colSpan: 8,
						editorOptions: { maxLength: 250, showClearButton: true },
					},
				],
			},
			{
				itemType: 'group',
				caption: 'Posición propuesta',
				colCount: 8,
				colSpan: 8,
				items: [
					{
						dataField: 'GERENCIA_PROPUESTA',
						label: { text: 'Gerencia / Vicerrectoría / Facultad' },
						colSpan: 4,
						template: 'CORR_GERENCIA_PROPUESTALookup',
					},
					{
						dataField: 'CORR_UNIDAD_PROPUESTA',
						label: { text: 'Departamento / Unidad' },
						colSpan: 4,
						template: 'CORR_UNIDAD_PROPUESTALookup',
					},
					{
						dataField: 'CORR_PUESTO_PROPUESTO',
						label: { text: 'Cargo' },
						colSpan: 4,
						template: 'CORR_PUESTO_PROPUESTOLookup',
					},
					{
						dataField: 'SALARIO_PROPUESTO',
						label: { text: 'Salario mensual' },
						colSpan: 2,
						editorType: 'dxNumberBox',
						editorOptions: { format: '#,##0.00', min: 0, showClearButton: true },
					},
					{
						dataField: 'CORR_TIPO_MODALIDAD_PROPUESTA',
						label: { text: 'Modalidad de trabajo' },
						colSpan: 2,
						template: 'CORR_TIPO_MODALIDAD_PROPUESTALookup',
					},
					{
						dataField: 'HORARIO_PROPUESTO',
						label: { text: 'Horario de trabajo' },
						colSpan: 8,
						editorOptions: { maxLength: 250, showClearButton: true },
					},
				],
			},
			{
				itemType: 'group',
				caption: 'Justificación',
				colCount: 8,
				colSpan: 8,
				items: [
					{
						dataField: 'JUSTIFICACION',
						label: { text: 'Justificación' },
						colSpan: 8,
						editorType: 'dxTextArea',
						editorOptions: { height: 90, maxLength: 1000 },
					},
				],
			},
		];
	}
	//#endregion
}
